File: client-go/queen.go
package queen

import (
	"context"
	"fmt"
	"time"
)

// Queen is the main client for Queen MQ.
type Queen struct {
	httpClient    *HttpClient
	bufferManager *BufferManager
	admin         *Admin
}

// New creates a new Queen client.
// Config can be:
//   - string: single URL
//   - []string: multiple URLs
//   - ClientConfig: full configuration
func New(config interface{}) (*Queen, error) {
	var clientConfig ClientConfig

	switch c := config.(type) {
	case string:
		clientConfig = ClientConfig{URL: c}
	case []string:
		clientConfig = ClientConfig{URLs: c}
	case ClientConfig:
		clientConfig = c
	case *ClientConfig:
		if c != nil {
			clientConfig = *c
		}
	default:
		return nil, fmt.Errorf("invalid config type: expected string, []string, or ClientConfig")
	}

	// Apply defaults
	if clientConfig.EnableFailover == false && len(clientConfig.URLs) == 0 && clientConfig.URL == "" {
		clientConfig.EnableFailover = ClientDefaults.EnableFailover
	}

	httpClient, err := NewHttpClient(clientConfig)
	if err != nil {
		return nil, fmt.Errorf("failed to create HTTP client: %w", err)
	}

	return &Queen{
		httpClient:    httpClient,
		bufferManager: NewBufferManager(),
	}, nil
}

// Queue returns a QueueBuilder for the specified queue.
func (q *Queen) Queue(name string) *QueueBuilder {
	return NewQueueBuilder(q, name)
}

// Admin returns the Admin API client.
func (q *Queen) Admin() *Admin {
	if q.admin == nil {
		q.admin = NewAdmin(q.httpClient)
	}
	return q.admin
}

// Transaction returns a new TransactionBuilder.
func (q *Queen) Transaction() *TransactionBuilder {
	return NewTransactionBuilder(q.httpClient)
}

// Ack acknowledges one or more messages.
// messages can be *Message, []*Message, or []Message
// success: true for completed, false for failed
func (q *Queen) Ack(ctx context.Context, messages interface{}, success bool, opts AckOptions) ([]AckResponse, error) {
	// Normalize messages to slice
	var msgs []*Message
	switch m := messages.(type) {
	case *Message:
		msgs = []*Message{m}
	case []*Message:
		msgs = m
	case []Message:
		for i := range m {
			msgs = append(msgs, &m[i])
		}
	case Message:
		msgs = []*Message{&m}
	default:
		return nil, fmt.Errorf("invalid messages type: expected *Message, []*Message, or []Message")
	}

	if len(msgs) == 0 {
		return nil, fmt.Errorf("at least one message is required")
	}

	// Validate messages
	if err := ValidateMessages(msgs); err != nil {
		return nil, err
	}

	// Determine status
	status := AckStatusCompleted
	if !success {
		status = AckStatusFailed
	}

	// Single message ack
	if len(msgs) == 1 {
		msg := msgs[0]
		req := ackRequest{
			TransactionID: msg.TransactionID,
			PartitionID:   msg.PartitionID,
			LeaseID:       msg.LeaseID,
			Status:        status,
			Error:         opts.Error,
			ConsumerGroup: opts.ConsumerGroup,
		}

		result, err := q.httpClient.Post(ctx, "/api/v1/ack", req)
		if err != nil {
			return nil, fmt.Errorf("ack request failed: %w", err)
		}

		// Parse response
		response := AckResponse{Success: true}
		if errMsg, ok := result["error"].(string); ok && errMsg != "" {
			response.Success = false
			response.Error = errMsg
		}

		logDebug("Queen.Ack", map[string]interface{}{
			"transactionId": msg.TransactionID,
			"status":        status,
			"success":       response.Success,
		})

		return []AckResponse{response}, nil
	}

	// Batch ack
	acks := make([]ackRequest, len(msgs))
	for i, msg := range msgs {
		acks[i] = ackRequest{
			TransactionID: msg.TransactionID,
			PartitionID:   msg.PartitionID,
			LeaseID:       msg.LeaseID,
			Status:        status,
			Error:         opts.Error,
		}
	}

	req := batchAckRequest{
		Acknowledgments: acks,
		ConsumerGroup:   opts.ConsumerGroup,
	}

	result, err := q.httpClient.Post(ctx, "/api/v1/ack/batch", req)
	if err != nil {
		return nil, fmt.Errorf("batch ack request failed: %w", err)
	}

	// Parse response
	var responses []AckResponse
	if data, ok := result["data"].([]interface{}); ok {
		for _, item := range data {
			if itemMap, ok := item.(map[string]interface{}); ok {
				resp := AckResponse{Success: true}
				if errMsg, ok := itemMap["error"].(string); ok && errMsg != "" {
					resp.Success = false
					resp.Error = errMsg
				}
				responses = append(responses, resp)
			}
		}
	} else {
		// Single response for batch
		resp := AckResponse{Success: true}
		if errMsg, ok := result["error"].(string); ok && errMsg != "" {
			resp.Success = false
			resp.Error = errMsg
		}
		for range msgs {
			responses = append(responses, resp)
		}
	}

	logDebug("Queen.Ack", map[string]interface{}{
		"count":   len(msgs),
		"status":  status,
		"success": len(responses),
	})

	return responses, nil
}

// Renew renews the lease for one or more messages.
// messages can be *Message, []*Message, []Message, string (leaseId), or []string (leaseIds)
func (q *Queen) Renew(ctx context.Context, messages interface{}) ([]RenewResponse, error) {
	// Normalize to lease IDs
	var leaseIDs []string
	switch m := messages.(type) {
	case string:
		leaseIDs = []string{m}
	case []string:
		leaseIDs = m
	case *Message:
		if m.LeaseID == "" {
			return nil, fmt.Errorf("message has no leaseId")
		}
		leaseIDs = []string{m.LeaseID}
	case []*Message:
		for _, msg := range m {
			if msg.LeaseID != "" {
				leaseIDs = append(leaseIDs, msg.LeaseID)
			}
		}
	case []Message:
		for _, msg := range m {
			if msg.LeaseID != "" {
				leaseIDs = append(leaseIDs, msg.LeaseID)
			}
		}
	case Message:
		if m.LeaseID == "" {
			return nil, fmt.Errorf("message has no leaseId")
		}
		leaseIDs = []string{m.LeaseID}
	default:
		return nil, fmt.Errorf("invalid messages type")
	}

	if len(leaseIDs) == 0 {
		return nil, fmt.Errorf("no lease IDs to renew")
	}

	// Renew each lease
	var responses []RenewResponse
	for _, leaseID := range leaseIDs {
		path := fmt.Sprintf("/api/v1/lease/%s/extend", leaseID)
		result, err := q.httpClient.Post(ctx, path, map[string]interface{}{})

		resp := RenewResponse{LeaseID: leaseID}
		if err != nil {
			resp.Success = false
			resp.Error = err.Error()
		} else {
			resp.Success = true
			if expiresAt, ok := result["newExpiresAt"].(string); ok {
				if t, err := time.Parse(time.RFC3339, expiresAt); err == nil {
					resp.NewExpiresAt = t
				}
			}
		}
		responses = append(responses, resp)

		logDebug("Queen.Renew", map[string]interface{}{
			"leaseId": leaseID,
			"success": resp.Success,
		})
	}

	return responses, nil
}

// FlushAllBuffers flushes all message buffers.
func (q *Queen) FlushAllBuffers(ctx context.Context) error {
	return q.bufferManager.FlushAll(ctx)
}

// GetBufferStats returns statistics about message buffers.
func (q *Queen) GetBufferStats() BufferStats {
	return q.bufferManager.GetStats()
}

// DeleteConsumerGroup deletes a consumer group.
func (q *Queen) DeleteConsumerGroup(ctx context.Context, consumerGroup string, deleteMetadata bool) error {
	path := fmt.Sprintf("/api/v1/consumer-groups/%s?deleteMetadata=%t", consumerGroup, deleteMetadata)
	_, err := q.httpClient.Delete(ctx, path)
	if err != nil {
		return fmt.Errorf("failed to delete consumer group: %w", err)
	}

	logInfo("Queen.DeleteConsumerGroup", map[string]interface{}{
		"consumerGroup":  consumerGroup,
		"deleteMetadata": deleteMetadata,
	})

	return nil
}

// UpdateConsumerGroupTimestamp updates the subscription timestamp for a consumer group.
func (q *Queen) UpdateConsumerGroupTimestamp(ctx context.Context, consumerGroup string, timestamp time.Time) error {
	path := fmt.Sprintf("/api/v1/consumer-groups/%s/subscription", consumerGroup)
	body := map[string]interface{}{
		"subscriptionTimestamp": timestamp.Format(time.RFC3339),
	}

	_, err := q.httpClient.Post(ctx, path, body)
	if err != nil {
		return fmt.Errorf("failed to update consumer group timestamp: %w", err)
	}

	logInfo("Queen.UpdateConsumerGroupTimestamp", map[string]interface{}{
		"consumerGroup": consumerGroup,
		"timestamp":     timestamp,
	})

	return nil
}

// Stream returns a StreamBuilder for defining a stream.
func (q *Queen) Stream(name, namespace string) *StreamBuilder {
	return NewStreamBuilder(q.httpClient, name, namespace)
}

// Consumer returns a StreamConsumer for consuming from a stream.
func (q *Queen) Consumer(streamName, consumerGroup string) *StreamConsumer {
	return NewStreamConsumer(q.httpClient, streamName, consumerGroup)
}

// Close gracefully shuts down the Queen client.
// It flushes all buffers before closing.
func (q *Queen) Close(ctx context.Context) error {
	logInfo("Queen.Close", map[string]interface{}{
		"status": "closing",
	})

	// Flush all buffers
	if err := q.bufferManager.FlushAll(ctx); err != nil {
		logError("Queen.Close", map[string]interface{}{
			"error": err.Error(),
		})
		return fmt.Errorf("failed to flush buffers: %w", err)
	}

	// Close HTTP client
	q.httpClient.Close()

	logInfo("Queen.Close", map[string]interface{}{
		"status": "closed",
	})

	return nil
}

// GetHttpClient returns the HTTP client (for internal use).
func (q *Queen) GetHttpClient() *HttpClient {
	return q.httpClient
}

// GetBufferManager returns the buffer manager (for internal use).
func (q *Queen) GetBufferManager() *BufferManager {
	return q.bufferManager
}

File: client-js/src/Queen.ts
/**
 * Queen Message Queue Client - Version 2
 * Clean, fluent API with smart defaults
 */

import { HttpClient } from './http/HttpClient'
import { LoadBalancer } from './http/LoadBalancer'
import { BufferManager } from './buffer/BufferManager'
import { QueueBuilder } from './builders/QueueBuilder'
import { TransactionBuilder } from './builders/TransactionBuilder'
import { StreamBuilder } from './stream/StreamBuilder'
import { StreamConsumer } from './stream/StreamConsumer'
import { Admin } from './admin/Admin'
import { CLIENT_DEFAULTS } from './utils/defaults'
import { validateUrl, validateUrls } from './utils/validation'
import * as logger from './utils/logger'
import type { QueenConfig, NormalizedQueenConfig, QueenMessage, AckContext, AckResult, RenewResult, BufferStats } from './types'

export class Queen {
  private httpClient: HttpClient
  private bufferManager: BufferManager
  private _config: NormalizedQueenConfig
  private shutdownHandlers: (() => void)[] = []
  private _admin: Admin | null = null

  constructor(config: QueenConfig | string | string[] = {}) {
    logger.log('Queen.constructor', { config: typeof config === 'object' && !Array.isArray(config) ? { ...config, urls: config.urls?.length || 0 } : { type: typeof config } })

    this._config = this.normalizeConfig(config)

    this.httpClient = this.createHttpClient()

    this.bufferManager = new BufferManager(this.httpClient)

    this.setupGracefulShutdown()

    logger.log('Queen.constructor', { status: 'initialized', urls: this._config.urls.length })
  }

  private normalizeConfig(config: QueenConfig | string | string[]): NormalizedQueenConfig {
    if (typeof config === 'string') {
      return {
        ...CLIENT_DEFAULTS,
        urls: [validateUrl(config)]
      }
    }

    if (Array.isArray(config)) {
      return {
        ...CLIENT_DEFAULTS,
        urls: validateUrls(config)
      }
    }

    const normalized = {
      ...CLIENT_DEFAULTS,
      ...config
    } as NormalizedQueenConfig & { url?: string }

    if (normalized.urls) {
      normalized.urls = validateUrls(normalized.urls)
    } else if (normalized.url) {
      normalized.urls = [validateUrl(normalized.url)]
    } else {
      throw new Error('Must provide urls or url in configuration')
    }

    delete normalized.url

    return normalized as NormalizedQueenConfig
  }

  private createHttpClient(): HttpClient {
    const { urls, timeoutMillis, retryAttempts, retryDelayMillis, loadBalancingStrategy, affinityHashRing, healthRetryAfterMillis, enableFailover, bearerToken, headers } = this._config

    if (urls.length === 1) {
      return new HttpClient({
        baseUrl: urls[0],
        timeoutMillis,
        retryAttempts,
        retryDelayMillis,
        bearerToken,
        headers
      })
    }

    const loadBalancer = new LoadBalancer(urls, loadBalancingStrategy, {
      affinityHashRing,
      healthRetryAfterMillis
    })
    return new HttpClient({
      loadBalancer,
      timeoutMillis,
      retryAttempts,
      retryDelayMillis,
      enableFailover,
      bearerToken,
      headers
    })
  }

  private setupGracefulShutdown(): void {
    if (typeof process === 'undefined' || typeof process.on !== 'function') {
      return
    }

    let signalReceivedCount = 0
    const shutdown = async (signal: string) => {
      console.log(`\nReceived ${signal}, shutting down gracefully...`)
      try {
        signalReceivedCount++
        if (signalReceivedCount > 1) {
          console.log('Received multiple shutdown signals, exiting immediately')
          process.exit(1)
        }
        await this.close()
        process.exit(0)
      } catch (error) {
        console.error('Error during shutdown:', error)
        process.exit(1)
      }
    }

    const sigintHandler = () => shutdown('SIGINT')
    const sigtermHandler = () => shutdown('SIGTERM')

    process.on('SIGINT', sigintHandler)
    process.on('SIGTERM', sigtermHandler)

    this.shutdownHandlers.push(
      () => process.off('SIGINT', sigintHandler),
      () => process.off('SIGTERM', sigtermHandler)
    )
  }

  // ===========================
  // Queue Builder Entry Point
  // ===========================

  queue(name: string | null = null): QueueBuilder {
    return new QueueBuilder(this, this.httpClient, this.bufferManager, name)
  }

  // ===========================
  // Admin API Entry Point
  // ===========================

  get admin(): Admin {
    if (!this._admin) {
      this._admin = new Admin(this.httpClient)
    }
    return this._admin
  }

  // ===========================
  // Transaction API
  // ===========================

  transaction(): TransactionBuilder {
    return new TransactionBuilder(this.httpClient)
  }

  // ===========================
  // Direct ACK API
  // ===========================

  async ack(message: QueenMessage | QueenMessage[] | string | string[], status: boolean | string = true, context: AckContext = {}): Promise<AckResult> {
    const isBatch = Array.isArray(message)
    logger.log('Queen.ack', { isBatch, count: isBatch ? message.length : 1, status, context })

    if (Array.isArray(message)) {
      if (message.length === 0) {
        return { processed: 0, results: [], success: true }
      }

      const hasIndividualStatus = message.some(msg =>
        typeof msg === 'object' && msg !== null && ('_status' in msg || '_error' in msg)
      )

      let acknowledgments: Array<{
        transactionId: string
        partitionId: string
        status: string
        error: string | null
        leaseId?: string
      }>

      if (hasIndividualStatus) {
        acknowledgments = message.map(msg => {
          const transactionId = typeof msg === 'string' ? msg : ((msg as QueenMessage).transactionId || (msg as QueenMessage).id)
          const partitionId = typeof msg === 'object' ? (msg as QueenMessage).partitionId : null
          const leaseId = typeof msg === 'object' ? (msg as QueenMessage).leaseId : null

          if (!transactionId) {
            throw new Error('Message must have transactionId or id property')
          }

          if (!partitionId) {
            throw new Error('Message must have partitionId property to ensure message uniqueness')
          }

          const msgStatus = (msg as QueenMessage)._status !== undefined ? (msg as QueenMessage)._status : status
          const statusStr = typeof msgStatus === 'boolean'
            ? (msgStatus ? 'completed' : 'failed')
            : msgStatus as string

          const ack: { transactionId: string; partitionId: string; status: string; error: string | null; leaseId?: string } = {
            transactionId,
            partitionId,
            status: statusStr,
            error: (msg as QueenMessage)._error || context.error || null
          }

          if (leaseId) ack.leaseId = leaseId

          return ack
        })
      } else {
        const statusStr = typeof status === 'boolean'
          ? (status ? 'completed' : 'failed')
          : status

        acknowledgments = message.map(msg => {
          const transactionId = typeof msg === 'string' ? msg : ((msg as QueenMessage).transactionId || (msg as QueenMessage).id)
          const partitionId = typeof msg === 'object' ? (msg as QueenMessage).partitionId : null
          const leaseId = typeof msg === 'object' ? (msg as QueenMessage).leaseId : null

          if (!transactionId) {
            throw new Error('Message must have transactionId or id property')
          }

          if (!partitionId) {
            throw new Error('Message must have partitionId property to ensure message uniqueness')
          }

          const ack: { transactionId: string; partitionId: string; status: string; error: string | null; leaseId?: string } = {
            transactionId,
            partitionId,
            status: statusStr,
            error: context.error || null
          }

          if (leaseId) ack.leaseId = leaseId

          return ack
        })
      }

      try {
        const result = await this.httpClient.post('/api/v1/ack/batch', {
          acknowledgments,
          consumerGroup: context.group || null
        }) as AckResult | null

        if (result && result.error) {
          logger.error('Queen.ack', { type: 'batch', error: result.error })
          return { success: false, error: result.error }
        }

        logger.log('Queen.ack', { type: 'batch', success: true, count: acknowledgments.length })
        return { success: true, ...result }
      } catch (error) {
        logger.error('Queen.ack', { type: 'batch', error: (error as Error).message })
        return { success: false, error: (error as Error).message }
      }
    }

    // Handle single message acknowledgment
    const transactionId = typeof message === 'string' ? message : ((message as QueenMessage).transactionId || (message as QueenMessage).id)
    const partitionId = typeof message === 'object' ? (message as QueenMessage).partitionId : null
    const leaseId = typeof message === 'object' ? (message as QueenMessage).leaseId : null

    if (!transactionId) {
      return { success: false, error: 'Message must have transactionId or id property' }
    }

    if (!partitionId) {
      return { success: false, error: 'Message must have partitionId property to ensure message uniqueness' }
    }

    const statusStr = typeof status === 'boolean'
      ? (status ? 'completed' : 'failed')
      : status

    const body: { transactionId: string; partitionId: string; status: string; error: string | null; consumerGroup: string | null; leaseId?: string } = {
      transactionId,
      partitionId,
      status: statusStr,
      error: context.error || null,
      consumerGroup: context.group || null
    }

    if (leaseId) body.leaseId = leaseId

    try {
      const result = await this.httpClient.post('/api/v1/ack', body) as AckResult | null

      if (result && result.error) {
        logger.error('Queen.ack', { type: 'single', transactionId, error: result.error })
        return { success: false, error: result.error }
      }

      logger.log('Queen.ack', { type: 'single', transactionId, success: true })
      return { success: true, ...result }
    } catch (error) {
      logger.error('Queen.ack', { type: 'single', transactionId, error: (error as Error).message })
      return { success: false, error: (error as Error).message }
    }
  }

  // ===========================
  // Lease Renewal API
  // ===========================

  async renew(messageOrLeaseId: string | QueenMessage | (string | QueenMessage)[]): Promise<RenewResult | RenewResult[]> {
    let leaseIds: string[] = []

    if (typeof messageOrLeaseId === 'string') {
      leaseIds = [messageOrLeaseId]
    } else if (Array.isArray(messageOrLeaseId)) {
      leaseIds = messageOrLeaseId.map(item =>
        typeof item === 'string' ? item : (item as QueenMessage).leaseId!
      ).filter(Boolean)
    } else if (messageOrLeaseId && typeof messageOrLeaseId === 'object') {
      if ((messageOrLeaseId as QueenMessage).leaseId) {
        leaseIds = [(messageOrLeaseId as QueenMessage).leaseId!]
      }
    }

    if (leaseIds.length === 0) {
      logger.warn('Queen.renew', 'No valid lease IDs found for renewal')
      return { leaseId: '', success: false, error: 'No valid lease IDs found for renewal' }
    }

    logger.log('Queen.renew', { count: leaseIds.length })

    const results: RenewResult[] = []
    for (const leaseId of leaseIds) {
      try {
        const result = await this.httpClient.post(`/api/v1/lease/${leaseId}/extend`, {}) as { leaseId?: string; newExpiresAt?: string; lease_expires_at?: string }
        results.push({
          leaseId,
          success: true,
          newExpiresAt: result.leaseId ? result.newExpiresAt : result.lease_expires_at
        })
        logger.log('Queen.renew', { leaseId, success: true })
      } catch (error) {
        results.push({ leaseId, success: false, error: (error as Error).message })
        logger.error('Queen.renew', { leaseId, error: (error as Error).message })
      }
    }

    logger.log('Queen.renew', { total: results.length, successful: results.filter(r => r.success).length })
    return Array.isArray(messageOrLeaseId) ? results : results[0]
  }

  // ===========================
  // Buffer Management API
  // ===========================

  async flushAllBuffers(): Promise<void> {
    logger.log('Queen.flushAllBuffers', 'Starting flush of all buffers')
    await this.bufferManager.flushAllBuffers()
    logger.log('Queen.flushAllBuffers', 'Completed')
  }

  getBufferStats(): BufferStats {
    const stats = this.bufferManager.getStats()
    logger.log('Queen.getBufferStats', stats)
    return stats
  }

  // ===========================
  // Consumer Group Management
  // ===========================

  async deleteConsumerGroup(consumerGroup: string, deleteMetadata: boolean = true): Promise<unknown> {
    logger.log('Queen.deleteConsumerGroup', { consumerGroup, deleteMetadata })

    const url = `/api/v1/consumer-groups/${encodeURIComponent(consumerGroup)}?deleteMetadata=${deleteMetadata}`
    const response = await this.httpClient.delete(url)

    logger.log('Queen.deleteConsumerGroup', { success: true, consumerGroup })
    return response
  }

  async updateConsumerGroupTimestamp(consumerGroup: string, timestamp: string): Promise<unknown> {
    logger.log('Queen.updateConsumerGroupTimestamp', { consumerGroup, timestamp })

    const url = `/api/v1/consumer-groups/${encodeURIComponent(consumerGroup)}/subscription`
    const response = await this.httpClient.post(url, {
      subscriptionTimestamp: timestamp
    })

    logger.log('Queen.updateConsumerGroupTimestamp', { success: true, consumerGroup })
    return response
  }

  // ===========================
  // Streaming API
  // ===========================

  stream(name: string, namespace?: string): StreamBuilder {
    logger.log('Queen.stream', { name, namespace })
    return new StreamBuilder(this.httpClient, this, name, namespace)
  }

  consumer(streamName: string, consumerGroup: string): StreamConsumer {
    logger.log('Queen.consumer', { streamName, consumerGroup })
    return new StreamConsumer(this.httpClient, this, streamName, consumerGroup)
  }

  // ===========================
  // Graceful Shutdown
  // ===========================

  async close(): Promise<void> {
    logger.log('Queen.close', 'Starting shutdown')
    console.log('Closing Queen client...')

    try {
      await this.bufferManager.flushAllBuffers()
      logger.log('Queen.close', 'All buffers flushed')
      console.log('All buffers flushed')
    } catch (error) {
      logger.error('Queen.close', { error: (error as Error).message, phase: 'buffer-flush' })
      console.warn('Error flushing buffers:', error)
    }

    this.bufferManager.cleanup()

    for (const cleanup of this.shutdownHandlers) {
      cleanup()
    }
    this.shutdownHandlers = []

    logger.log('Queen.close', 'Client closed successfully')
    console.log('Queen client closed')
  }
}
